Extract reward before observation (C++ and Python)

gasse opened the pull request:
See #209

gasse commented:
I reordered reward, observation everywhere to be consistent (even in the constructor where it does not matter).

The tests seem to pass, at least on my machine.

@AntoinePrv can you check the code for C++ (great)ness ?

gasse commented:
All CI tests pass now. Good to go ?

AntoinePrv commented:
There is a long commit history that does not seem to belong to this PR.

gasse commented:
Oh no, I though I had merged that one before the release :O

@AntoinePrv that's because I have rebased dev onto master. I've rebased this PR as well, should look better now.

gasse commented:
I guess v0.7.0 will have a very short life...

dchetelat commented on libecole/include/ecole/environment/environment.hpp:
```diff
@@ -112,22 +112,30 @@ class Environment {
-			}
+			// Extract reward, observation and information (in that order)
+			auto&& reward = reward_function().extract(model(), done);
+			auto&& observation = [&, done = done]() -> OptionalObservation {
```
It's fine I guess, but isn't it a bit overkill to use an anonymous function here? Why not just a ternary operator `= ? :`

AntoinePrv replied:
Agreed, instantly invoked lambdas are only useful to initialize `const` variables. Also `done` is redudant with `&`.

gasse replied:
I agree it is a bit of an overkill, but I tried both and nonetheless I find this more readable. With ` = ? : ` you have to specify the class `OptionalObservation` twice, plus our code formatter does a terrible job at splitting the code on two lines. I let you judge:
```diff
-			auto&& observation = [&, done = done]() -> OptionalObservation {
-                               if (done) {  // Don't extract observations in final states
-                                       return {};
-                               }
-                               return observation_function().extract(model(), done);
-                       }();
+                       auto&& observation = done ? OptionalObservation{} :  // Don't extract observations in final states
+                                                                                                                OptionalObservation{observation_function().extract(model(), done)};
```

dchetelat commented on libecole/include/ecole/environment/environment.hpp:
```diff
@@ -170,19 +178,26 @@ class Environment {
-				observation = observation_function().extract(model(), done);
-			}
+			// Extract reward, observation and information (in that order)
+			auto&& reward = reward_function().extract(model(), done);
+			auto&& observation = [&, done = done]() -> OptionalObservation {
```
Same point here

AntoinePrv replied:
```suggestion
			auto reward = reward_function().extract(model(), done);
			auto observation = [&, done = done]() -> OptionalObservation {
```

AntoinePrv commented on libecole/include/ecole/environment/environment.hpp:
```diff
@@ -112,22 +112,30 @@ class Environment {
-			auto const [done, action_set] = dynamics().reset_dynamics(model(), std::forward<Args>(args)...);
+
+			// Place the environment in its initial state
+			auto&& [done, action_set] = dynamics().reset_dynamics(model(), std::forward<Args>(args)...);
```
```suggestion
			auto [done, action_set] = dynamics().reset_dynamics(model(), std::forward<Args>(args)...);
```

AntoinePrv commented on libecole/include/ecole/environment/environment.hpp:
```diff
@@ -112,22 +112,30 @@ class Environment {
-				observation = observation_function().extract(model(), done);
-			}
+			// Extract reward, observation and information (in that order)
+			auto&& reward = reward_function().extract(model(), done);
+			auto&& observation = [&, done = done]() -> OptionalObservation {
```
```suggestion
			auto reward = reward_function().extract(model(), done);
			auto observation = [&, done = done]() -> OptionalObservation {
```

AntoinePrv commented on libecole/include/ecole/environment/environment.hpp:
```diff
@@ -112,22 +112,30 @@ class Environment {
+				}
+				return observation_function().extract(model(), done);
+			}();
+			auto&& information = information_function().extract(model(), done);
```
```suggestion
			auto information = information_function().extract(model(), done);
```

gasse replied:
This is true for gcc, but for clang `done=done` is required, otherwise it finds our code is not valid. Something to do with `done` being a structured binding. See:
https://stackoverflow.com/questions/46114214/lambda-implicit-capture-fails-with-variable-declared-from-structured-binding

AntoinePrv commented on libecole/include/ecole/environment/environment.hpp:
```diff
@@ -112,22 +112,30 @@ class Environment {
-				std::move(observation),
-				std::move(action_set),
-				reward_function().extract(model(), done),
+				observation,
+				action_set,
+				reward,
```
```suggestion
				std::move(observation),
				std::move(action_set),
				std::move(reward),
```

AntoinePrv commented on libecole/include/ecole/environment/environment.hpp:
```diff
@@ -112,22 +112,30 @@ class Environment {
+				reward,
 				done,
-				information_function().extract(model(), done),
+				information,
```
```suggestion
				std::move(information),
```

AntoinePrv commented on libecole/include/ecole/environment/environment.hpp:
```diff
@@ -170,19 +178,26 @@ class Environment {
 		try {
-			auto const [done, action_set] = dynamics().step_dynamics(model(), action, std::forward<Args>(args)...);
+			// Transition the environment to the next state
+			auto&& [done, action_set] = dynamics().step_dynamics(model(), action, std::forward<Args>(args)...);
```
```suggestion
			auto [done, action_set] = dynamics().step_dynamics(model(), action, std::forward<Args>(args)...);
```

AntoinePrv commented on libecole/include/ecole/environment/environment.hpp:
```diff
@@ -170,19 +178,26 @@ class Environment {
+				}
+				return observation_function().extract(model(), done);
+			}();
+			auto&& information = information_function().extract(model(), done);
```
```suggestion
			auto information = information_function().extract(model(), done);
```

AntoinePrv commented on libecole/include/ecole/environment/environment.hpp:
```diff
@@ -170,19 +178,26 @@ class Environment {
-				std::move(observation),
-				std::move(action_set),
-				reward_function().extract(model(), done),
+				observation,
+				action_set,
+				reward,
```
```suggestion
				std::move(observation),
				std::move(action_set),
				std::move(reward),
```

AntoinePrv commented on libecole/include/ecole/environment/environment.hpp:
```diff
@@ -170,19 +178,26 @@ class Environment {
+				reward,
 				done,
-				information_function().extract(model(), done),
+				information,
```
```suggestion
				std::move(information),
```

gasse commented:
Thanks for the code review ! I'll make the changes ASAP.

gasse commented:
I removed the `auto&&` universal pointer stuff, switched back to `std::move()` in returns, and changed the lambda for a ternary operator. I also put some stuff inside a private function, to avoid a bit of code duplication.

@dchetelat @AntoinePrv let me know how that looks to you now.

AntoinePrv commented on libecole/include/ecole/environment/environment.hpp:
```diff
@@ -170,19 +171,19 @@ class Environment {
 		try {
-			auto const [done, action_set] = dynamics().step_dynamics(model(), action, std::forward<Args>(args)...);
+			// Transition the environment to the next state
+			auto&& [done, action_set] = dynamics().step_dynamics(model(), action, std::forward<Args>(args)...);
```
```suggestion
			auto [done, action_set] = dynamics().step_dynamics(model(), action, std::forward<Args>(args)...);
```

gasse commented:
Ah it looks like some conflicting commits have not been handled correctly when I rebased `dev` onto `master` for v0.7.0 ... Thanks @dchetelat for noticing that !

dchetelat commented:
Yeah I feel it's the third time we fix that lol. Rebasing chaos...
Also I rewrote again the ternary thing, the compiler and the code check looks happy with it. All good to go for me as well!
